docs/index.md

```diff
@@ -296,6 +296,10 @@ describe('#find()', function () {
 });
 ```
 
+### Limitations of asynchronous callbacks
+
+You can use all three asynchronous patterns (`done`, `Promise`, and `async`/`await`) in callbacks for `it()` and for all hooks (`before()`, `after()`, `beforeEach()`, `afterEach()`), but `describe()` will not work correctly with an asynchronous callback -- it must be synchronous. This is because in the "parsing" cycle all `describe` callbacks are executed and the test hierarchy (`runner`) is created before any tests are run.
```

Comment:
```suggestion
You can use all asynchronous callbacks (`done`, `Promise`, and `async`/`await`) in callbacks for `it()`, `before()`, `after()`, `beforeEach()`, `afterEach()`) but not `describe()` -- it must be synchronous.
```

Reply:
OK for all changes except that I feel like understanding the 'why' really helped me here.

Reply:
Yeah, I'm thinking its maybe getting a bit too much into the current implementation of things, I'll have a think about it.

If we include it then maybe as a new paragraph, like `Technically this is caused by...`? Thoughts @mochajs/maintenance-crew?

Reply:
Yeah, and it can always be found by looking in the GitHub history of that documentation line as it will reference this PR 😌

Reply:
I am still keen on including a technical reason -- it's a fundamental design in mocha that is seriously not obvious, and it took me (and others, read the linked issues) a _long_ time to find out why my tests weren't working properly. (You can always remove the explanation in the unlikely event of mocha redesign making it unimportant...)

IMO the docs are for technical readers; they shouldn't be dumbed down to the point of 'do this because we tell you to'. That's really not helpful for users; they end up using mocha without understanding it, and get things wrong -- like this.

Little details in technical documentation are the things that make a project stand out.

But if you really don't want that extra detail, I'll pull it out.

Reply:
How about:

```suggestion
You can use all asynchronous callbacks (`done`, `Promise`, and `async`/`await`) in callbacks for `it()`, `before()`, `after()`, `beforeEach()`, `afterEach()`) but not `describe()` -- it must be synchronous.
See [#5046](https://github.com/mochajs/mocha/pull/5046) for more information.
```

> IMO the docs are for technical readers

I think the difference here isn't from whether the docs are for technical readers, it's to what extent readers want those technical bits from high-level docs. Many readers on a top-level page like https://mochajs.org are just seeking the _"do this because we tell you to"_. For those users, the choice is to either give that to them, or have them zone out and find something else.

Ideally we should have deeper docs to link to that explain things like architecture and design choices. I'd be very in favor of a site reorg that encompasses that. That'd let us have both the surface information _and_ deeper bits. +1 to the urge to give users the option to actually understand Mocha!

Reply:
> Ideally we should have deeper docs to link to that explain things like architecture and design choices. I'd be very in favor of a site reorg that encompasses that.

+1 on this, agree that would be the ideal.

Happy with your suggestion -- strikes a good compromise for those who want to understand why.